src/HDF5.jl

```diff
@@ -1468,10 +1453,10 @@ function write(obj::DatasetOrAttribute, data::VLen{T}) where {T<:Union{ScalarTyp
     end
 end
 # For plain files and groups, let "write(obj, name, val; properties...)" mean "d_write"
-write(parent::Union{File,Group}, name::String, data::Union{T,AbstractArray{T}}; pv...) where {T<:Union{ScalarType,String,Complex{<:ScalarType}}} =
+write(parent::Union{File,Group}, name::AbstractString, data::Union{T,AbstractArray{T}}; pv...) where {T<:Union{ScalarType,String,Complex{<:ScalarType}}} =
     d_write(parent, name, data; pv...)
 # For datasets, "write(dset, name, val; properties...)" means "a_write"
-write(parent::Dataset, name::String, data::Union{T, AbstractArray{T}}; pv...) where {T<:ScalarType,String} =
+write(parent::Dataset, name::AbstractString, data::Union{T,AbstractArray{T}}; pv...) where {T<:Union{ScalarType,String}} =
```

Comment:
Good point, there's something wrong here. On this PR `which` is pointing to the `deprecated` methods... arrg

Reply:
This is a bit strange and I'm not sure how to go about it (actually unrelated to this PR, so could be follow up work), but still need to address writing AbstractStrings
```julia
julia> @which write(d,"A","A")
write(parent::HDF5.Dataset, name::AbstractString, data::Union{AbstractArray{T, N} where N, T, StridedMatrix{T}, StridedVector{T}, StridedArray{T, N} where N, BitMatrix, BitVector}, plists::HDF5.Properties...) where T<:Union{Bool, Float32, Float64, Int16, Int32, Int64, Int8, UInt16, UInt32, UInt64, UInt8, HDF5.Reference, String} in HDF5 at C:\Users\Mus\.julia\dev\HDF5\src\deprecated.jl:144

julia> @which write(d,"A","A";)
(::Base.var"#write##kw")(::Any, ::typeof(write), parent::HDF5.Dataset, name::AbstractString, data::Union{AbstractArray{T, N} where N, T, StridedMatrix{T}, StridedVector{T}, StridedArray{T, N} where N, BitMatrix, BitVector}) where T<:(Union{Bool, Float32, Float64, Int16, Int32, Int64, Int8, UInt16, UInt32, UInt64, UInt8, HDF5.Reference, var"#s27"} where var"#s27"<:AbstractString) in HDF5 at C:\Users\Mus\.julia\dev\HDF5\src\HDF5.jl:1459

```

Reply:
Edit, ignore that last comment.... user error (was modifying some things)

Reply:
Sorry, on my branch what I get is the following:
```julia
julia> which(write, (HDF5.Dataset, String, String))
write(parent::HDF5.Dataset, name::AbstractString, data::Union{AbstractArray{T, N} where N, T, StridedMatrix{T}, StridedVector{T}, StridedArray{T, N} where N, BitMatrix, BitVector}; pv...) where T<:Union{Bool, Float32, Float64, Int16, Int32, Int64, Int8, UInt16, UInt32, UInt64, UInt8, HDF5.Reference, String} in HDF5 at C:\Users\Mus\.julia\dev\HDF5\src\HDF5.jl:1459

julia> which(write, (HDF5.Dataset, String, String))^C

julia> which(write, (HDF5.Dataset, String, Int))
write(parent::HDF5.Dataset, name::AbstractString, data::Union{AbstractArray{T, N} where N, T, StridedMatrix{T}, StridedVector{T}, StridedArray{T, N} where N, BitMatrix, BitVector}; pv...) where T<:Union{Bool, Float32, Float64, Int16, Int32, Int64, Int8, UInt16, UInt32, UInt64, UInt8, HDF5.Reference, String} in HDF5 at C:\Users\Mus\.julia\dev\HDF5\src\HDF5.jl:1459

```

So they are calling the correct methods.  I just need to fix them to allow writing `AbstractString`

Reply:
Does it make sense to allow 
```
write(parent::Dataset, name::AbstractString, data::Union{T,AbstractArray{T}}; pv...) where {T<:Union{ScalarType,<:AbstractString}}
```
?

In general this would require custom user methods to permit the conversion. I dont' necessarily see anything wrong with widening the signature, in fact I think this is the correct direction to go. 

I also noticed that we don't have `o_delete` methods for ` o_delete(::HDF5.Attribute)` , so when it errors it throws an exception to the exception (lol)